fml/message_loop_impl.cc

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/fml/message_loop_impl.h"
+#include "lib/ftl/build_config.h"
+
+#if OS_MACOSX
+
+#include "flutter/fml/platform/darwin/message_loop_darwin.h"
+using PlatformMessageLoopImpl = fml::MessageLoopDarwin;
+
+#else
+
+#error This platform does not have a message loop implementation.
+
+#endif
+
+namespace fml {
+
+ftl::RefPtr<MessageLoopImpl> MessageLoopImpl::Create() {
+  return ftl::MakeRefCounted<::PlatformMessageLoopImpl>();
+}
+
+MessageLoopImpl::MessageLoopImpl() : order_(0) {}
+
+MessageLoopImpl::~MessageLoopImpl() = default;
+
+void MessageLoopImpl::PostTask(ftl::Closure task, ftl::TimePoint target_time) {
+  if (task == nullptr) {
+    return;
+  }
+
+  WakeUp(RegisterTaskAndGetNextWake(task, target_time));
+}
+
+void MessageLoopImpl::RunExpiredTasksNow() {
+  WakeUp(RunExpiredTasksAndGetNextWake());
+}
+
+void MessageLoopImpl::SetTaskObserver(MessageLoop::TaskObserver observer) {
+  ftl::MutexLocker lock(&task_observer_mutex_);
```

Comment:
You shouldn't need a lock for this.  It should be an error to call this function on a thread other than the loop's own thread.  Maybe DCHECK that the current message loop has |this| for its impl?